lib/Differentiator/ReverseModeVisitor.cpp

```diff
@@ -75,7 +89,12 @@ namespace clad {
     Expr* PopExpr =
         m_Sema.ActOnCallExpr(getCurrentScope(), PopDRE, noLoc, TapeRef, noLoc)
             .get();
-    Expr* CallArgs[] = {TapeRef, E};
+    Expr* exprToPush = E;
+    if (auto AT = dyn_cast<ArrayType>(E->getType())) {
```

Comment:
```suggestion
    if (isa<ArrayType>(E->getType())) {
```
We will need to add an assertion in `getArraySizeExpr` if that was not an array type; and cast the type there.

Reply:
Why? It adds two unnecessary checks instead of just one. Secondly, even if I do what you suggest I see no scenario in which ```getArraySizeExpr``` will see a non ```ArrayType```.

Reply:
This way we reduce the cyclomatic code complexity and as for the size you can see the suggestion in https://github.com/vgvassilev/clad/pull/513/files#r1057265867

Reply:
Where are these two checks? If you mean that we will have a check outside of the function and then inside the function implementation then sure. However, the way I'd read this code is "hey we have a special case for arrays" and it's being taken care of in this utility function. Leaking a variable also makes the reader track what goes on in the utility function and outside - that's not to worry today but will have to worry tomorrow.

Reply:
I disagree here. I think the dynamic cast provides a clear-er message and avoids redundant checks. However, we can move the cast inside if you like.

Reply:
If that's an unambiguous code readability improvement for you then let's keep it as it is.

Reply:
Okay, thanks!